### PR TITLE
New MVA-based Tau-Ids (update)

### PR DESCRIPTION
**Pull request by mbluj**
The MVA-based Tau-IDs are retrained with 94X Fall17 samples and should be evaluated using MiniAOD Taus (skimmedTaus) and added to NanoAOD. This PR is meant to supersede its initial version (#108) based on MVA training with 92X Summer17 samples.

The commits change three files:
 * RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py with definition of MVA payloads (new ones are added),
 * PhysicsTools/NanoAOD/python/taus_cff.py where new Tau-Ids are evaluated and added to the Tau table
 * PhysicsTools/NanoAOD/python/nanoDQM_cfi.py

Notes: 
* This PR bases on CMSSW_9_4_5_cand1 and master of NanoAOD repository.
* Implementation in the PR is expected to give same results on MiniAOD v1 and v2 thanks to employing run2_miniAOD_94XFall17 era for v2. However, it crashes with v1 inputs due to some electron dependences
* It is considered to add two int numbers to code WPs of two cut-based tau-Ids as requested by end users.

FYI, @veelken @steggema @ohlushch

### CODE REVIEW

**Comment by veelken** [comment]
Dear all,

I see that the two int numbers for the cut-based tau ID have not been added to PhysicsTools/NanoAOD/python/taus_cff.py . Can these please still be added ? I think the option to switch to the cut-based tau ID discriminators may be a useful option in case of problems with data/MC differences in the MVA tau ID input variables (the Tau POG is seeing some evidence for such data/MC differences right now and is investigating).

File: PhysicsTools/NanoAOD/python/taus_cff.py [hunk 1]
```diff
@@ -3,13 +3,314 @@
...
+    src = cms.InputTag("slimmedTausUpdated"),
+    cut = cms.string("pt > 18 && tauID('decayModeFindingNewDMs') && (tauID('byLooseCombinedIsolationDeltaBetaCorr3Hits') || tauID('byVLooseIsolationMVArun2v1DBoldDMwLT') || tauID('byVLooseIsolationMVArun2v1DBnewDMwLT') || tauID('byVLooseIsolationMVArun2v1DBdR03oldDMwLT') || tauID('byVVLooseIsolationMVArun2v1DBoldDMwLT2017v2') || tauID('byVVLooseIsolationMVArun2v1DBnewDMwLT2017v2') || tauID('byVVLooseIsolationMVArun2v1DBdR03oldDMwLT2017v2'))")
+)
+run2_miniAOD_94XFall17.toModify(finalTaus,
+          cut =  cms.string("pt > 18 && tauID('decayModeFindingNewDMs') && (tauID('byLooseCombinedIsolationDeltaBetaCorr3Hits') || tauID('byVLooseIsolationMVArun2v1DBoldDMwLT2015') || tauID('byVLooseIsolationMVArun2v1DBnewDMwLT') || tauID('byVLooseIsolationMVArun2v1DBdR03oldDMwLT') || tauID('byVVLooseIsolationMVArun2v1DBoldDMwLT2017v2') || tauID('byVVLooseIsolationMVArun2v1DBnewDMwLT2017v2') || tauID('byVVLooseIsolationMVArun2v1DBdR03oldDMwLT2017v2'))")
 )
```
Comment (gpetruc) [inline PhysicsTools/NanoAOD/python/taus_cff.py]: don't use `run2_miniAOD_94XFall17` era
Have the default config to be that for 94XMiniAODv2, and use `run2_nanoAOD_94XMiniAODv1` to adapt to MiniAOD v1
(I'm putting just one comment but it is true for all uses of this era in the file)

File: RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py [hunk 2]
```diff
@@ -41,6 +41,8 @@
 }
```
Comment (gpetruc) [inline RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py]: diff too is not being too smart here.
I would assume this only loads new items compared to what's currently in the release, can you confirm?

**Comment by steggema** [comment]
Hi Michal,
Thanks for the fast turnaround! We agreed yesterday to support both the 2017 MC v1-based and 2017 MC v2-based trainings. Do I read correctly that this PR adds the v2-based training, plus reverts to the 2015-based training for NanoAOD production based on MiniAOD v2? I think it would be better to offer the 2017 MC v1-based training, either in addition or replacing the 2015 training.

File: PhysicsTools/NanoAOD/python/taus_cff.py [hunk 3]
```diff
@@ -3,13 +3,314 @@
...
+)
+_tauIDSourcesExt = cms.PSet(
+    slimmedTausUpdated.tauIDSources,
+    _tauIDSources2015
+)
+run2_miniAOD_94XFall17.toModify(slimmedTausUpdated,
+          tauIDSources = _tauIDSourcesExt
+)
+
+patTauMVAIDsSeq += slimmedTausUpdated
```
Comment (gpetruc) [inline PhysicsTools/NanoAOD/python/taus_cff.py]: can you put all this in a separate cff file (e.g. taus_updatedMVAIds_cff) ?

File: PhysicsTools/NanoAOD/python/taus_cff.py [hunk 4]
```diff
@@ -3,13 +3,314 @@
...
+# MVAIso DBoldDM Seqeunce
+patTauDiscriminationByIsolationMVArun2v1DBoldDMwLT2015Seq = cms.Sequence(
+    patTauDiscriminationByIsolationMVArun2v1DBoldDMwLTraw2015
+    + patTauDiscriminationByVLooseIsolationMVArun2v1DBoldDMwLT2015
+    + patTauDiscriminationByLooseIsolationMVArun2v1DBoldDMwLT2015
+    + patTauDiscriminationByMediumIsolationMVArun2v1DBoldDMwLT2015
+    + patTauDiscriminationByTightIsolationMVArun2v1DBoldDMwLT2015
+    + patTauDiscriminationByVTightIsolationMVArun2v1DBoldDMwLT2015
+    + patTauDiscriminationByVVTightIsolationMVArun2v1DBoldDMwLT2015
+)
```
Comment (gpetruc) [inline PhysicsTools/NanoAOD/python/taus_cff.py]: is 2015 what is in current miniAOD v1?

**Review by gpetruc — CHANGES_REQUESTED** [review]
* use run2_nanoAOD_94XMiniAODv1/v2 eras for 94X
 * it's not clear what's happening on 80X

**Comment by mbluj** [comment]
Hi

On Tue, Mar 27, 2018 at 4:31 PM, Christian Veelken <notifications@github.com
> wrote:

> Dear all,
>
> I see that the two int numbers for the cut-based tau ID have not been
> added to PhysicsTools/NanoAOD/python/taus_cff.py . Can these please still
> be added ? I think the option to switch to the cut-based tau ID
> discriminators may be a useful option in case of problems with data/MC
> differences in the MVA tau ID input variables (the Tau POG is seeing some
> evidence for such data/MC differences right now and is investigating).
>
It is in plans, i.e. I will add it as soon as I can (hopefully by tomorrow
about noon).




<#m_-5256863754617431948_DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by mbluj** [comment]
On Tue, Mar 27, 2018 at 4:46 PM, Jan Steggemann <notifications@github.com>
wrote:

> Hi Michal,
> Thanks for the fast turnaround! We agreed yesterday to support both the
> 2017 MC v1-based and 2017 MC v2-based trainings. Do I read correctly that
> this PR adds the v2-based training, plus reverts to the 2015-based training
> for NanoAOD production based on MiniAOD v2? I think it would be better to
> offer the 2017 MC v1-based training, either in addition or replacing the
> 2015 training.
>
OK, it can be done like this, i.e. 2017v1 can be added. The issue with the
2017v1 is that it exists only for oldDMs dR=0.5 (however it is most used
version).

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

File: PhysicsTools/NanoAOD/python/taus_cff.py (hunk 1 above)
Comment (mbluj) [reply PhysicsTools/NanoAOD/python/taus_cff.py]: OK, thanks for letting me know, I will correct it.

File: RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py (hunk 2 above)
Comment (mbluj) [reply RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py]: Yes, I confirm that it only adds new 2017v2 payloads. In particular, it means that standard RECO and MiniAOD v1 and v2 workflows are not broken.

File: PhysicsTools/NanoAOD/python/taus_cff.py (hunk 3 above)
Comment (mbluj) [reply PhysicsTools/NanoAOD/python/taus_cff.py]: OK.

File: PhysicsTools/NanoAOD/python/taus_cff.py (hunk 4 above)
Comment (mbluj) [reply PhysicsTools/NanoAOD/python/taus_cff.py]: Yes, exactly

**Comment by mbluj** [comment]
Dear All,
the two last commits address comments on:
* moving definitions of MVA tau-Ids to a separate cff file
* usage of correct eras (hopefully it is fine now)
* adding MVAIso 2017v1 which is supported by Tau POG.

It does not address, however, the request to add cut-based tau-Ids as due to some reasons there is a problem with parsing of mathematical expressions like: "tauID('byCombinedIsolationDeltaBetaCorrRaw3Hits')<1.5" ("<" sign looks be a problem). 
Advises how to cope with this problem are welcome.

**Comment by gpetruc** [comment]
Hi @mbluj,
Where do you need to parse the `x < 1.5` ?
If it's in the table producer, then comparisons should be supported for variables of type `bool`, as the physicstools parser grammar allows them in cuts but not in expressions.
If you need it in an expression, e.g. to say `1*(x<1.5) + 2*(y>4.7)` then you need to use the ternary operator but with one extra question mark at the beginning (because of limitations in the parser of the grammar), so the above would become `(? x<1.5 ? 1 : 0) +  (? y>4.7 ? 2 : 0)`

**Comment by mbluj** [comment]
Hi,

On Wed, Mar 28, 2018 at 1:24 PM, Giovanni Petrucciani <
notifications@github.com> wrote:

> Hi @mbluj <https://github.com/mbluj>,
> Where do you need to parse the x < 1.5 ?
> If it's in the table producer, then comparisons should be supported for
> variables of type bool, as the physicstools parser grammar allows them in
> cuts but not in expressions.
> If you need it in an expression, e.g. to say 1*(x<1.5) + 2*(y>4.7) then
> you need to use the ternary operator but with one extra question mark at
> the beginning (because of limitations in the parser of the grammar), so the
> above would become (? x<1.5 ? 1 : 0) + (? y>4.7 ? 2 : 0)
>
Yes, it is expected to go to the table producer. I have checked that it
indeed works for a single bool, but what I want is to store a few of bools
(a set of WPs) in one integer (uint8), like this (a simplified example):
idCombIsodR03 =  Var(
            "1 * (tauID('chargedIsoPtSumdR03') < 2.5) + "
            "2 * (tauID('chargedIsoPtSumdR03') < 1.5) ",
            "uint8", doc="Combined isolation WP"
)
I will try to adapt to your proposal, but it looks that it will produce a
quite complicated expression :(

**Comment by arizzi** [comment]
please use the proper Era Switches to support MC campaigns
94X MiniAOD v1
94X MiniAOD v2
80X MiniAOD
and 92X , 80X for data

**Comment by mbluj** [comment]
It looks that 80X jobs fail. How they should be setup? Same as MiniAODv1 ones? Should I use a specific era?

**Comment by arizzi** [comment]
there are specific eras for the various scenarios
https://github.com/cms-sw/cmssw/blob/master/Configuration/StandardSequences/python/Eras.py#L47-L48

**Comment by mbluj** [comment]
On Wed, Mar 28, 2018 at 3:55 PM, arizzi <notifications@github.com> wrote:

> there are specific eras for the various scenarios
> https://github.com/cms-sw/cmssw/blob/master/Configuration/
> StandardSequences/python/Eras.py#L47-L48
>
Thank you, I know this. But while meaning and use cases for '
run2_miniAOD_94XFall17', 'run2_nanoAOD_92X' and
'run2_nanoAOD_92X' are more or less clear I am not sure what should be used
for 80X as there is not specific era for nanoAOD with 80X. And it makes
difference if it will be 'run2_miniAOD_80XLegacy' or something else.

**Comment by mbluj** [comment]
Now eras should correctly handle 94X MiniAOD v1 and v1, 92X MiniAOD, and 80X MiniAOD and LegacyMiniAOD.

Concerning cut-based isolation tau-Ids: 
* WPs for dR=0.5 were trivially added as they are present in MiniAOD, 
* while for dR=0.3 one has to use a set ternary operators to implement logic of cuts. I have tried several possible implementations (one of which is commented out in git repo), but although the implementations technically work they do not return sensible results. It could be that the expressions are too complicated by the parser.

P.S. I'm away since ~5pm today and will be able provide only reduced fixes to the PR until end of Easter.

**Comment by veelken** [comment]
Hi Michal,

sorry, I do not see the need for computing the cut-based tau ID WP for dR=0.3 with a complex expression. Why can we not simply recompute the cut-based tau ID WP for the slimmedTaus collection, like you do it for the MVA-based discriminators ? I thought we had a plugin to recompute the cut-based WPs in RecoTauTag/RecoTau/plugins/ already (loading the different isolation pT sums from InputTags).

**Comment by mbluj** [comment]
Hi

On Wed, Mar 28, 2018 at 5:11 PM, Christian Veelken <notifications@github.com
> wrote:

> Hi Michal,
>
> sorry, I do not see the need for computing the cut-based tau ID WP for
> dR=0.3 with a complex expression. Why can we not simply recompute the
> cut-based tau ID WP for the slimmedTaus collection, like you do it for the
> MVA-based discriminators ? I thought we had a plugin to recompute the
> cut-based WPs in RecoTauTag/RecoTau/plugins/ already (loading the different
> isolation pT sums from InputTags).
>
I'm sorry but I'm not sure if such plugin exists. If it is the case, please
point me to it and I'll be happy to use it here.

**Comment by mbluj** [comment]
Hello,
I have checked again and I do not see the plugin mentioned by @veelken. So, I see two possible quick solutions:
* one will make string-parser-based solution working or
* we will store raw (delta-beta corrected) isolation sum with dR=0.3 to allow end users to compute WPs themselves. Note that pt-photon-outside-cone, which is 2nd component of cut-based isolation, is already stored. It will however require to store a new float (precision=10 should be fine). 
If 2nd solution is acceptable we can remove WPs for cut based isolation with dR=0.5 (uint8) as it can be computed in the same way. We can also consider to remove some other variables stored as floats, but it will require clear decision from conveners of Tau POG and probably also of HtoTaus.

**Comment by veelken** [comment]
Hi,

I looked for a plugin that can compute the WP using isolation pT sums as input and I also do not find it. I recall that the RECO conveners asked for such a plugin and I remember that Pavel has been working on it, but it seems Pavel couldn't finish it before he left CMS. 

Michal, I have think it is possible to simplify the string expression to (i.e. no need to use the "?" syntax):
    idCombIsodR03 = Var( "1*( (tauID('photonPtSumOutsideSignalConedR03')<(0.1*pt)) && ((tauID('chargedIsoPtSumdR03')+max(0.,tauID('neutralIsoPtSumdR03')-0.072*tauID('puCorrPtSum')))<2.5) ) + 2*( (tauID('photonPtSumOutsideSignalConedR03')<(0.1*pt)) && ((tauID('chargedIsoPtSumdR03')+max(0.,tauID('neutralIsoPtSumdR03')-0.072*tauID('puCorrPtSum')))<1.5) ) + 4*( (tauID('photonPtSumOutsideSignalConedR03')<(0.1*pt)) && ((tauID('chargedIsoPtSumdR03')+max(0.,tauID('neutralIsoPtSumdR03')-0.072*tauID('puCorrPtSum')))<0.8) )")

I could write this plugin in 1-2 days, in case that it sufficient in terms of timescale and you could test it.

In my opinion, the 2nd solution of storing the raw (delta-beta corrected) isolation sum for dR=0.3 would be ok too.

**Comment by gpetruc** [comment]
@veelken as discussed above, comparison operators are not supported without the "?" syntax in variables.
I don't see any problem with the multiline string that is now commented out in the code, it's long but is readable.

In general, I'm a bit perplexed by the large number of IDs one would end up with this PR - not from the perspective of disk size (I haven't checked that yet) but of continued support all these for analysis. What is the long term plan here, e.g. for summer analyses? 
It may be fine to temporarily provide also old IDs (especially on miniAOD v1), but in that case it should be clearly advertised to the analyses, so that they all plan towards using the new ones and we can remove the old ones in a subsequent nanoAOD production.

**Comment by steggema** [comment]
@gpetruc The Tau POG plans to generally recommend the latest MVA-base discriminators using old decay mode finding, i.e.
  idMVAoldDM2017v1
  idMVAoldDM2017v2
  idMVAoldDMdR032017v2 (for analyses in dense environments; current sole customer is ttH)

We also still support cut-based discriminators as a simpler-to-understand backup. While we don't formally plan to support the 2015-based training (idMVAoldDM), we observe larger differences from unity in the scale factors for the 2017-based training, so this is more of a backup solution in case this will turn out to outweigh the nominal advantages of the new training. 

We also support that analyses test whether new decay mode finding improves their analysis, however, we haven't seen a single case where this is the case in run 2 (it is still used by at least one analysis).

From the purpose of tau POG work, I personally find it more useful to store the inputs to the cut-based discriminators rather than the decisions of the IDs, since the latter can be easily recalculated; however, if the additional storage need of the latter is negligible, there's nothing wrong with including them.

**Comment by mbluj** [comment]
Hello,

On Thu, Mar 29, 2018 at 10:21 AM, Giovanni Petrucciani <
notifications@github.com> wrote:

> @veelken <https://github.com/veelken> as discussed above, comparison
> operators are not supported without the "?" syntax in variables.
> I don't see any problem with the multiline string that is now commented
> out in the code, it's long but is readable.
>
It is readable, but due to some reasons does not return what is expected...
So either there is a problem with the expression (which I tried to test
independently) or the string parser has a problem with it. As written
above, I tried a couple logically equal expressions but they returned
incorrect results (different in each case).

**Comment by mbluj** [comment]
Hello,

On Thu, Mar 29, 2018 at 11:45 AM, Jan Steggemann <notifications@github.com>
wrote:

> @gpetruc <https://github.com/gpetruc> The Tau POG plans to generally
> recommend the latest MVA-base discriminators using old decay mode finding,
> i.e.
> idMVAoldDM2017v1
> idMVAoldDM2017v2
> idMVAoldDMdR032017v2 (for analyses in dense environments; current sole
> customer is ttH)
>
> We also still support cut-based discriminators as a simpler-to-understand
> backup. While we don't formally plan to support the 2015-based training
> (idMVAoldDM), we observe larger differences from unity in the scale factors
> for the 2017-based training, so this is more of a backup solution in case
> this will turn out to outweigh the nominal advantages of the new training.
>
> We also support that analyses test whether new decay mode finding improves
> their analysis, however, we haven't seen a single case where this is the
> case in run 2 (it is still used by at least one analysis).
>
> From the purpose of tau POG work, I personally find it more useful to
> store the inputs to the cut-based discriminators rather than the decisions
> of the IDs, since the latter can be easily recalculated; however, if the
> additional storage need of the latter is negligible, there's nothing wrong
> with including them.
>
Taking the above under consideration, my proposal is as follows:
 * Keep WPs of MVAIso tau-Ids with the old 2015 MC based training (three
uint8 numbers), but remove their raw values (three float numbers)
 * Add WPs of  MVAIso tau-Ids with the 2017 MC v1 based training and
corresponding raw value (one uint8 and one float number)
 * Add WPs of  MVAIso tau-Ids with the 2017 MC v1 based training and
corresponding raw value (three uint8 and three float numbers)
 * Add raw delta-beta corrected sum with dR=03 (one float number)
 * Do not add WPs for cut based isolation as they can be easy recalculated
from raw isolation sums.
 * Remove "footprintCorr" (a float number)

In total it gives:
 * one additional float number (precision=10)
 * four additional uint8 numbers

If it sounds acceptable for all parties, I will try to update the PR along
these lines by this evening.

**Comment by gpetruc** [comment]
ok to me

**Comment by veelken** [comment]
Sounds good to me too.

Jan, I assume you mean to store the raw MVA output for the new 2017 v2 training, i.e.
* Add WPs of MVAIso tau-Ids with the 2017 MC v2 based training and corresponding raw value (three uint8 and three float numbers)
right ?

**Comment by mbluj** [comment]
On Tue, Apr 3, 2018 at 11:45 AM, Christian Veelken <notifications@github.com
> wrote:

> Sounds good to me too.
>
> Jan, I assume you mean to store the raw MVA output for the new 2017 v2
> training, i.e.
>
>    - Add WPs of MVAIso tau-Ids with the 2017 MC v2 based training and
>    corresponding raw value (three uint8 and three float numbers)
>    right ?
>
> Yes, exactly it concerns raw MVAIso score values. The raw values are used
to order tau candidates before a given WP is used to actually select an
event. For 2017v2 training there are three raw values for: old DMs dR=0.5,
new DMs dR=0.5 and old DMs dR=0.3.

I'm going to proceed with updating the PR.

Michał

**Comment by mbluj** [comment]
A set of last commits is a result of a discussion with tau and h->taus experts and consists of:
* removal of rarely used MVAIso Tau-IDs (WPs and raw outputs) with 2015 training (two uint8 and two float numbers)
* removal of one isolation variable "footprint correction" (one float number)
* adding combined isolation sum for cone of dR=0.3 (one float number)
* adding MVAIso Tau-IDs (WPs and raw outputs) with 2017 training v1 and v2 (four uint8 and four float numbers)

We would like to ask you to test the configuration.

**Review by gpetruc — APPROVED** [review]

**Comment by gpetruc** [comment]
I'm ok with the changes in 2017 data, but I'm perplexed by the switch to 2017 training also for 80X 2016 data / mc.
I think it would make sense to keep in 80X also the current trainings, as current analyses are based on the old ones.

**Comment by mbluj** [comment]
Observed differences look reasonable. For instance, most notable difference in the "Long test mc94Xv2" whose is marked as "passed with differences" is distribution of MVAIso oldDM score called "rawMVAoldDM" which corresponds with change of its input and is expected.

**Comment by mbluj** [comment]
Situation with 80X is a bit complicated, but hopefully under control:
 * we keep WPs and raw score for the most used MVAIso tau-ID called oldDM with 2015 training which are useful to repeat already published selections when NanoAOD build on original 80X MiniAOD samples,
 * we offer new MVAIso tau-IDs (2017 v1 and v2 training) for new analyses when NanoAOD build on top of ReMiniAOD of 2016 data/MC with 94X (which consist of tau re-reco).
It could be that one can solve this even better, but it will require to differentiate between the above two situations (at least in my naive understanding) => new eras?

**Comment by gpetruc** [comment]
the current 80X era is for the 80X processing of 2016 data, not the re-miniAOD of 2016 data with 94X (that one will have a new era, when the re-miniAOD will be made).
that's why I would keep the old discriminators; in particular I'd keep also the oldDMdR03 which is used in ttH/susy analyses.

**Comment by mbluj** [comment]
OK, it makes sense to me. So I will keep all 2015 trainings for Nano on top of 80X Mini, but not add the new ones as they are not compatible. I suppose that it will require era dependent DQM... 
BTW, configuration for Nano on top of (nonexistent) ReMiniAOD of 2016 samples will be similar as for 94X MiniAOD...

**Comment by mbluj** [comment]
The last commit to not rerun 2017v1/v1 MVAIso Tau-IDs and to keep all 2015 MVAIso Tau-IDs for Nano on top of 80X samples.

**Comment by gpetruc** [comment]
triggering tests again

**Comment by arizzi** [comment]
are the missing variables in DQM produced in some of the other eras? @gpetruc which era is the dqm tested on ?

**Comment by mbluj** [comment]
On Wed, Apr 4, 2018 at 8:50 AM, arizzi <notifications@github.com> wrote:

> are the missing variables in DQM produced in some of the other eras?
> @gpetruc <https://github.com/gpetruc> which era is the dqm tested on ?
>
Content of the tau table (other tables as well?) depends on era and it
could be indeed useful to make DQM also era depended. Anyway, I have
defined plots which make sense for all areas, i.e. there are histograms
which correspond with variables present in some eras but not in others
which triggers the warning by the DQM test.

**Comment by arizzi** [comment]
@mbluj  giovanni suggested that you store in the cfi only the 94X latest (i.e. default) config  and use the DQM_cff.py to apply era switches that could add back the legacy variables.
Can you try implementing it? (let me know if you can do it today, otherwise I may try to have a look at it at some point)

**Comment by mbluj** [comment]
On Wed, Apr 4, 2018 at 2:14 PM, arizzi <notifications@github.com> wrote:

> @mbluj <https://github.com/mbluj> giovanni suggested that you store in
> the cfi only the 94X latest (i.e. default) config and use the DQM_cff.py to
> apply era switches that could add back the legacy variables.
> Can you try implementing it? (let me know if you can do it today,
> otherwise I may try to have a look at it at some point)
>
OK, I can spend some time on it. I'll let you know if I fail.

**Comment by mbluj** [comment]
Era dependence added to NanoDQM_cff for tau plots. However, I have not idea how to test it properly...

**Comment by arizzi** [comment]
cherry picked on master-cmsswmaster and masterMergedAndNoData-integration

**Comment by arizzi** [comment]
https://github.com/cms-sw/cmssw/pull/22840

**Comment by vlimant** [comment]
@all what was the intention is pulling taggers from outside of the GT here ?

**Comment by mbluj** [comment]
Hello,
This was "temporary" solution to enable access to tauID payloads before they were available via GT (used around beginning of Run-2). Then it was used for testing purposes, etc. Anyway, up to my knowledge this was removed from CMSSW already some time ago (about ~2017?), but let me check. Moreover, tauIDs using these payloads are generally not supported for some time already.

**Comment by mbluj** [comment]
After further checks: there is still possibility to run old tauIDs (early Run-2) accessing their payloads w/o GT for certain Run-2 nanoAOD eras. This is steered by the following cff: `PhysicsTools/NanoAOD/python/taus_updatedMVAIds_cff.py`. The idea was to enable access to consistent set of versions of those IDs independently on version of miniAOD which was used to produce given nanoAOD. It is however not used for long time at least since Run2 legacy production (10_2?). I do not remember why it was kept for newer releases, probably was overlooked. It is not needed to read Run-2 UL samples (10_6) or even to reproduce most recent versions of the mentioned tauID on top of the UL samples as their payloads are accessible via GT.
I think I can prepare quickly a PR removing this stuff from NanoAOD sequences, i.e. remove `PhysicsTools/NanoAOD/python/taus_updatedMVAIds_cff.py`.

**Comment by vlimant** [comment]
thanks for checking, please prepare a PR indeed. 
N.B. we have workflows running from 10.6 MINI, modified with the era `run2_nanoAOD_106Xv2` that seem to still pull in the ID from outside the GT ; for these workflows, which we still want to maintain, the corresponding GT will have to be updated accordingly

**Comment by mbluj** [comment]
> thanks for checking, please prepare a PR indeed. N.B. we have workflows running from 10.6 MINI, modified with the era `run2_nanoAOD_106Xv2` that seem to still pull in the ID from outside the GT ; for these workflows, which we still want to maintain, the corresponding GT will have to be updated accordingly

I agree, it makes sense to maintain `run2_nanoAOD_106Xv2` era as it is the only way to produce Nano with UL samples with new releases. But, it is strange to me that old tauID is used by an official workflow with this era. I have to check this in details. Is thre another wf. (in matrix tests) affected by this issue than one reported by Kati?

**Pull request merged**